Avoid a duplicate declaration of openldap-clients package

werekraken opened the pull request:
Hopefully, it is not a breach of etiquette to open another PR. I had attempted a PR in chutzimir's fork https://github.com/chutzimir/puppet-openldap/pull/1.

This has the same goal as https://github.com/camptocamp/puppet-openldap/pull/114, to be able to have openldap::client and openldap::server on the same RHEL/CentOS system. This version passes all unit tests.

mcanevet commented on manifests/client/install.pp:
```diff
@@ -5,7 +5,19 @@
+      }
+    }
+    'RedHat': {
+      class { '::openldap::client::utils':
```
We may have to use `include ::openldap::client::utils` here to avoid duplicate declaration on `Class[openldap::client::utils]`

werekraken replied:
Using include will break the package parameter for openldap::client. This is why the unit tests fail in https://github.com/camptocamp/puppet-openldap/pull/114.

openldap::server already uses include, so there shouldn't be a duplicate declaration within the module. A duplicate declaration could occur if class is used outside the module (e.g. profiles, site.pp) to declare openldap::client and openldap::client::utils on a RedHat system. Avoiding that seems preferable to not being able to pass a custom the openldap::client::package in my opinion (especially since openldap::client exposes every current parameter for openldap::client::utils).

mcanevet replied:
@werekraken we may need to add a `$utils_package` parameter in `params`, expose it to `init` and use it as default in `openldap::client::utils` instead.

werekraken replied:
One issue I see with this is that currently there is no init.pp, but rather a client.pp and a server.pp. The issue that this PR is trying to address is that client.pp and server.pp both want to install the openldap-clients package on RedHat systems. That package resource needs to end up in the catalogue only once when openldap::client and openldap::server are both declared.

There are many ways we could address the duplicate resource. Some alternate solutions I can see are
1. Use ensure_packages/ensure_resource on the openldap-clients.
2. Make openldap::clients::utils not be included in server when on RedHat (or a parameter to bypass the include). It could still be declared directly via class/include, but would be mutually exclusive with openldap::client on RedHat.
3. Make openldap::client::utils completely decoupled from openldap::server. On Debian you could declare openldap::client, openldap::server, and openldap::client::utils as needed. On RedHat you could declare openldap::server and openldap::client or openldap::client::utils, but not both openldap::client and openldap::client::utils.
4. Expose an init.pp which could be used to configure the util_package. Have openldap::client include openldap::client::utils on RedHat. Have openldap::server include openldap::client::utils. Have openldap::client::utils use $openldap::utils_pacakge for the package resource.

I'm happy to code up an alternate implementation as time permits. Is adding an init.pp which provides a singular access to a parameter what you want?

mcanevet commented:
@werekraken hmm, `puppetlabs-postgresql` is designed a bit like your 4th proposal, but uses `globals` instead of `init`. This could be a good choice, if it does not require too much changes. Otherwise, we could use your 2nd proposal.

werekraken commented:
@mcanevet I looked into this a bit more today and made an interesting discovery.

Debian and RedHat both have 3 packages that are more or less equivalent. The previous complexities were based on the assumption that RedHat only had 2 packages, one of which joined the client and utils packages on Debian.

The equivalent packages are
| | Debian | RedHat |
| ------------- | ------------- | ------------- |
| "server" | slapd | openldap-servers |
| "client" | libldap-2.4-2 | openldap |
| "utils"  | ldap-utils | openldap-clients |

I moved the utils class to be top-level and added openldap::params::utils_package. The openldap::parmas::client_package was actually installing the openldap rpm as a dependency.

This implementation uses the Debian behavior for openldap::clients (which does not install tools like ldapadd, ldapmodify, etc). If we wanted to standardize on the RedHat previous behavior, I can easily add an include openldap::utils to openldap::client::install.

mcanevet commented:
@werekraken great! Is think we should still have an `openldap::client::utils` class that declares `openldap::utils` that throws a deprecation warning for compatibility. Could you please add it? Otherwise +1 to merge this.

werekraken commented:
@mcanevet Good idea. Added openldap::client::utils as requested.

mcanevet commented:
@werekraken LGTM, thanks
